### PR TITLE
Some improvements with types & duplicated code

### PR DESCRIPTION
**Pull request by Minigugus**
This Pull Request updates some points :

### `moreOptions` can be typed
https://github.com/rigwild/exoplatform-api-wrapper/blob/c980e4a086ac36ad7c6021ca2b39ad2d42d27744/src/index.ts#L38
### `method` is default `'GET'` but can be infered from the `body` parameter
https://github.com/rigwild/exoplatform-api-wrapper/blob/c980e4a086ac36ad7c6021ca2b39ad2d42d27744/src/index.ts#L37
### `body` is always `{}` if undefined, so `this.request('http://example.com')` will send make a `GET` HTTP request with a body
https://github.com/rigwild/exoplatform-api-wrapper/blob/c980e4a086ac36ad7c6021ca2b39ad2d42d27744/src/index.ts#L36
https://github.com/rigwild/exoplatform-api-wrapper/blob/c980e4a086ac36ad7c6021ca2b39ad2d42d27744/src/index.ts#L51-L54
https://github.com/rigwild/exoplatform-api-wrapper/blob/c980e4a086ac36ad7c6021ca2b39ad2d42d27744/src/index.ts#L76
### `postUser` & `postSpace` looks very similar and can be refactored to remove duplication
https://github.com/rigwild/exoplatform-api-wrapper/blob/c980e4a086ac36ad7c6021ca2b39ad2d42d27744/src/index.ts#L110-L116
https://github.com/rigwild/exoplatform-api-wrapper/blob/c980e4a086ac36ad7c6021ca2b39ad2d42d27744/src/index.ts#L126-L132

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -33,9 +33,9 @@ class ExoplatformBot {
    */
   request(
     path: string,
-    body: object = {},
-    method: string = 'GET',
-    moreOptions?: object
+    body: object | null = null,
+    method: string = body ? 'POST' : 'GET',
```
Comment (rigwild) [inline src/index.ts]: This is wrong as eXo Platform's API is REST. You can send a body with other HTTP methods, like PUT.

File: src/index.ts [hunk 2]
```diff
@@ -33,9 +33,9 @@ class ExoplatformBot {
...
+    moreOptions?: Partial<RequestOptions>
```
Comment (rigwild) [inline src/index.ts]: Nice find!

File: src/index.ts [hunk 3]
```diff
@@ -108,11 +120,7 @@ class ExoplatformBot {
    * @throws Unknown user or no permission to post
    */
   async postUser(userId: string, message: string): Promise<object | string> {
-    if (!this.username || !this.password) throw new Error(msgId.NEED_LOGGED_IN)
-
-    const uri = `/private/v1/social/users/${userId}/activities`
-    const res = await this.request(uri, { title: message }, 'POST')
-    return res.body
+    return (await this.loggedInRequest(`/private/v1/social/users/${userId}/activities`, { title: message })).body
```
Comment (rigwild) [inline src/index.ts]: Must fix this and bellow to select HTTP method.

File: src/index.ts [hunk 4]
```diff
@@ -99,6 +99,18 @@ class ExoplatformBot {
     }
   }
 
+  /**
+   * Make an API call to eXo Platform configured API, but throws if not logged in..
+   * @param args Same as `this.request`
+   * @returns Same as `this.request`
+   * @throws {Error} `this.login` must be called before `this.loggedInRequest`
+   */
+  private async loggedInRequest(...args: Parameters<ExoplatformBot['request']>): ReturnType<ExoplatformBot['request']> {
```
Comment (rigwild) [inline src/index.ts]: That is cool, didn't knew about typescript private fields. I tried using Node v12 new private field syntax (`#field`), but it didnt work.
Maybe `exoHostname`, `exoPath`, `exoSecureProtocol`, `username`, `password` should all be private too.

File: src/index.ts (hunk 1 above)
Comment (Minigugus) [reply src/index.ts]: It's just the *default* value :smile: The idea is just to remove the necessity of the `'POST'` here `this.request(yourUrl, yourBody, 'POST')`, but it's always possible to do so ^^

File: src/index.ts (hunk 4 above)
Comment (Minigugus) [reply src/index.ts]: I was thinking it was intended so I don't touched them :smile: 
`exoHostname`, `exoPath` and `exoSecureProtocol` may stay `public` as their are user-submitted parameters, in the constructor moreover :thinking: 
But it's sure that `username` & `password` should be `private` as they require validation in `login`, and are sensitives informations :lock:

Comment (rigwild) [reply src/index.ts]: I agree. Do these changes and I will merge it. Thanks!

Comment (Minigugus) [reply src/index.ts]: Done :smile: 
I just made use of some TypeScript magic
https://github.com/rigwild/exoplatform-api-wrapper/blob/4f8cf6b04e996c64b503d80952e9528cd797311b/src/index.ts#L16

**Pull request merged**